### PR TITLE
Feat/Sandbox

### PR DESCRIPTION
**Pull request by TheExGenesis**
For now, we have a sandboxed server that we can hit up from outside the sandbox with code and it will return a serialized walker object. (see `sandbox/server/index.py`
1. receive POST requests containing Python encodings for Walkers as strings, 
2. generate the Walker in the sandboxed server, 
3. return the serialized walker to the caller.

Evaluation can in principle be done outside of the sandbox since the arbitrary code is only used for generating walklers, which are then safe.

Open questions: 
- the example walker descriptions in python do generate walkers but then running `walker.validate()` returns false. I suspect it's a problem with the validate function rather than the descriptions but did not check.

### CODE REVIEW

**Comment by TheExGenesis** [comment]
Just realized I pulled from `build-lego` which isn't finished yet.

**Comment by honglu2875** [comment]
A note on the sandbox for my branch #13:
(my code is kind of quick-and-dirty, so feel free to do what's the best for your code)
If #13 is merged, for the toy task `ImageOptim` to run, I will need an api in the sandbox kind of like the following
```python
@app.route("/eval_func", methods=["POST"])
def evaluate_function():
    req_json = request.get_json()
    code_string = req_json["code"]
    func_name = req_json["func_name"]

    return execute_with_result(code_string, func_name, 5.0).__repr__()
```
where `execute_with_result` is just to 
- run the program under `reliability_guard()` in a process, 
- take in an extra string of function name (`func_name`) in the parameter, 
- execute the function, 
- assuming the result is a `np.ndarray` and return to me the `.tolist()` result. 

We just need to have a slightly different `unsafe_execute` kind of like the following
```
def unsafe_execute_with_result(check_program, func_name, result, timeout):
    import shutil, os
    reliability_guard()

    # Run program.
    try:
        exec_globals = {}
        with swallow_io():
            with time_limit(timeout):
                exec(check_program, exec_globals)
                result.append(exec_globals[func_name]().tolist())

    except BaseException as e:
        result.append(str(e))
```

**Comment by herbiebradley** [comment]
@TheExGenesis Thanks for the great work! Can you pull from main, make sure your code still works and can interface with @honglu2875's code (see https://github.com/CarperAI/ELM/blob/main/map_elites/environments.py#L213), then I think we're ready to merge!

**Comment by TheExGenesis** [comment]
@honglu2875 can I ask why the name `_with_result` and why you're appending to result rather than just returning?

**Comment by honglu2875** [comment]
> @honglu2875 can I ask why the name `_with_result` and why you're appending to result rather than just returning?
> 
> Also it seems to me different environments might benefit from different dedicated functions since e.g. your interface assumes an ndarray and uses the sandbox for a full eval, while the sodaracer only requires the sandbox to generate the racers, not for eval. I'll accommodate the ImageOptim of course.

Oh yeah I literally copied&pasted the other codes. The name `_with_result` is nothing important (I was thinking I'm evaluating the function and returning its result). Appending instead of returning is also unnecessary (simply because I copied from the other `unsafe_execute` function). My codes here are really ugly (plus a serious concurrency issue if we have multiple processes) and I only want to show an example of what the API needs to return. Please forget about them and only keep the idea that it should return the numpy array or list.

Yes, I agree that a different function and interface make total sense. Feel free to also change the name of the endpoint if necessary. The modification on my side is trivial.

I remember @herbiebradley had some codes somewhere about classifying different errors into different numbers. Maybe we can incorporate this into the sandbox and let it return json message like `{'status': ..., 'data': ...}` for both interface?

**Comment by TheExGenesis** [comment]
Thanks for responding! I'll make the error codes nicer by returning json. Don't have a great idea of what to do concurrency-wise for the execs, open to suggestions.

**Comment by TheExGenesis** [comment]
@herbiebradley confirmed that the code still works and interface is compatible with H's work.

File: sandbox/server/index.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+app = Flask(__name__)
+
+def bad_request(message, **kwargs):
+    return {'message': message, **kwargs}, 500
+
+
+def generate_racer(code_str, timeout):
+    try:
+        execution_result = sandbox_unsafe_execute(code_str, "make_walker", timeout)
```
Comment (herbiebradley) [inline sandbox/server/index.py:15]: I can't see `sandbox_unsafe_execute` in utils, only `unsafe_execute`?

File: sandbox/scripts/bootstrap.sh [hunk 2]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
```
Comment (herbiebradley) [inline sandbox/scripts/bootstrap.sh:1]: Is this script just for debugging?

File: sandbox/server/walker/walk_creator.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
+
```
Comment (herbiebradley) [inline sandbox/server/walker/walk_creator.py:1]: Can you update this file with the new `walk_creator` from the main branch? https://github.com/CarperAI/ELM/blob/main/walker/walk_creator.py

**Review by herbiebradley — COMMENTED** [review]
Looks good, I added a few comments.

File: sandbox/server/index.py (hunk 1 above)
Comment (TheExGenesis) [reply sandbox/server/index.py:15]: I was missing the rename!

File: sandbox/scripts/bootstrap.sh (hunk 2 above)
Comment (TheExGenesis) [reply sandbox/scripts/bootstrap.sh:1]: It's used in the Dockerfile to start the server

**Comment by TheExGenesis** [comment]
> Looks good, I added a few comments. Thanks for the work!

all good, handled the comments :)

**Review by herbiebradley — APPROVED** [review]
LGTM, I'll merge now!

**Pull request merged**